.doom.d/config.org
#+TITLE: DOOM Emacs Configuration
#+STARTUP: indent overview
#+PROPERTY: header-args :tangle config.el

* Look And Feel
** Theme
It's Catppuccin time!
#+begin_src emacs-lisp
  (use-package! catppuccin-theme
    :init (setq catppuccin-flavor 'macchiato)
    :config (setq doom-theme 'catppuccin))
#+end_src
** Default font
#+begin_src emacs-lisp
  (setq doom-font (font-spec :family "Inconsolata" :size 16 :weight 'medium))
#+end_src
** Working Hours
Helper methods for determining whether operating during regular working hours.
#+begin_src emacs-lisp
  (require 'ts)

  (defun my/is-working-hours ()
    (let ((now (ts-now)))
      (and (>= (ts-hour now) 8)
           (< (ts-hour now) 18)
           (>= (ts-dow now) 1)
           (<= (ts-dow now) 5))))
#+end_src
** Dashboard
Customize the Doom Emacs dashboard with commissioned artwork, chosen randomly.
The dashboard can be reloaded with the =g= key.

#+begin_src emacs-lisp
  (require 'f)

  (defun my/choice (things)
    (let ((index (random (length things))))
      (nth index things)))

  (defun my/dashboard-randomize ()
    (interactive)
    (let ((logo (my/choice
                 (seq-filter (lambda (img)
                               (and (f-exists? (plist-get img :img))
                                    (if (my/is-working-hours)
                                        (not (plist-get img :nsfw))
                                      't)))
                             '((:nsfw 't :img "~/Nextcloud/Pictures/Patreon/Jon Bliss/phoenix-and-catgirl-500.png")
                               (:nsfw 't :img "~/Nextcloud/Pictures/Patreon/Jon Bliss/Bassist-final-transparent-500.png")
                               (:nsfw 't :img "~/Nextcloud/Pictures/Patreon/Jon Bliss/catgirl-final-transparent-500.png")
                               (:nsfw 't :img "~/Nextcloud/Pictures/Patreon/Jon Bliss/lapiz-final-transparent-500.png")
                               (:nsfw 't :img "~/Nextcloud/Pictures/Patreon/Jon Bliss/Poledancer-final-transparent-500.png")
                               (:nsfw 't :img "~/Nextcloud/Pictures/Patreon/Jon Bliss/Hacker-final-transparent-500.png")
                               (:nsfw nil :img "~/Nextcloud/Pictures/Patreon/Bee/IMG_3666_transparent_500.png"))))))

      (setq fancy-splash-image (plist-get logo :img))
      (if (called-interactively-p)
          (+doom-dashboard-reload))))

  (my/dashboard-randomize)
  (map!
   :map +doom-dashboard-mode-map
   "g" #'my/dashboard-randomize)
#+end_src
** Frame transparency
Set the background transparency to 90% so my wallpaper shows through.

#+begin_src emacs-lisp
  (set-frame-parameter (selected-frame) 'alpha '(98 . 98))
  (add-to-list 'default-frame-alist '(alpha . (98 . 98)))
#+end_src
** Comic Sans Mono
Why not, it's fun 🤣

Adds a command to switch to the monospaced Comic Sans font, and make it the
default font for PHP buffers.

https://dtinth.github.io/comic-mono-font/

#+begin_src emacs-lisp
  (defun my/buffer-face-comic-mono ()
    (interactive)
    (setq buffer-face-mode-face '(:family "Comic Mono" :weight 'medium))
    (buffer-face-mode))

  (after! php-mode
    (add-hook 'php-mode-hook #'my/buffer-face-comic-mono))
#+end_src
** Emoji Support
Use the Noto Emoji font.
#+begin_src emacs-lisp
  (setf use-default-font-for-symbols nil)
  (set-fontset-font t 'unicode "Noto Emoji" nil 'append)
#+end_src
** Mixed Pitch
Facilitates mixing monospace and proportional fonts. I'm using an MIT-licensed
version of the [[https://github.com/edwardtufte/et-book][Edward Tufte book font]] because it's /gorgeous/.
#+begin_src emacs-lisp
  (use-package! mixed-pitch
    :hook ((org-mode markdown-mode rst-mode) . mixed-pitch-mode))
  (setq doom-variable-pitch-font (font-spec :family "ETBookOT" :weight 'regular))
#+end_src
* UX Improvements
** Disable C-z backgrounding when the GUI is loaded

In OSX, this minimizes the window. It's incredibly annoying, and I accidentally
hit it too much. I also can't think of any reason I'd want to background Emacs
outside of a terminal session.

#+begin_src emacs-lisp
  (if (display-graphic-p)
      (global-unset-key (kbd "C-z")))
#+end_src
** Custom settings
Store options set via =customize-*= in a separate file (Emacs stores
them in =init.el= by default).

#+BEGIN_SRC emacs-lisp
  (setq custom-file "~/.emacs.d/custom.el")
  (if (file-exists-p custom-file)
      (load custom-file))

  (setf custom-safe-themes t)
#+END_SRC
** Eval and Replace
Replace an s-expression with its value. Taken from [[http://emacsredux.com/blog/2013/06/21/eval-and-replace/][Emacs Redux]].

#+begin_src emacs-lisp
  (defun eval-and-replace ()
    "Replace the preceding sexp with its value."
    (interactive)
    (backward-kill-sexp)
    (condition-case nil
        (prin1 (eval (read (current-kill 0)))
               (current-buffer))
      (error (message "Invalid expression")
             (insert (current-kill 0)))))

  (global-set-key (kbd "C-)") 'eval-and-replace)
#+end_src
** Prompt for unsafe local variables
Doom sets this to =:safe=, logging unsafe variables for later addressing. I'd
rather continue to be prompted.

#+begin_src emacs-lisp
  (setq enable-local-variables t)
#+end_src
** Environment loading
Supports ~.envrc~ files as well as Python tools like Hatch, Poetry, and PDM.

#+begin_src emacs-lisp
  (use-package buffer-env
    :commands (buffer-env-update buffer-env-reset))
#+end_src
* Writing

** BibTeX
Tell Emacs where to find my bibliography files. I keep most everything in my
Calibre library, which I regularly export and keep synced via Nextcloud.

#+begin_src emacs-lisp
  (setq my/bibliographies
        '("~/Documents/bibliography/references.bib"
          "~/Documents/bibliography/calibre.bib"))
#+end_src
*** Helm BibTeX
#+begin_src emacs-lisp
  (setq bibtex-completion-bibliography my/bibliographies
        bibtex-completion-pdf-field "File"
        bibtex-completion-notes-path "~/org/roam")
#+end_src
*** Citar
#+begin_src emacs-lisp
  (setq citar-bibliography my/bibliographies
        citar-notes-paths '("~/org/roam"))
#+end_src
** Org
:PROPERTIES:
:header-args: :tangle no :noweb-ref org
:END:

Configure a variety of options and tools for [[https://orgmode.org][Org Mode]], the markup I use for
everything from simple notes to task management.

#+begin_src emacs-lisp :noweb yes :tangle "config.el" :noweb-ref org-all
  (after! org
    <<org>>)
#+end_src

*** Override DOOM indentation behavior
#+begin_src emacs-lisp
  (defun my/org-init-babel ()
    (setq org-src-preserve-indentation nil))

  (add-hook! 'org-mode-hook #'my/org-init-babel)
#+end_src
*** Disable DOOM's centralized attachment system
It's incompatible with all of the org files I already have using the standard
setup.
#+begin_src emacs-lisp
  (setq org-attach-directory "data/")
  (remove-hook! 'org-load-hook
    #'(+org-init-centralized-attachments-h))
#+end_src
*** Agenda
Set up my agenda view. I use separate files for my personal TODOs and my work
TODOs, synced externally using Nextcloud.

#+begin_src emacs-lisp
  (after! org-agenda
    (require 'f)

    (setq my/agenda-files '((personal . ("~/Nextcloud/org/personal.org")))

          org-agenda-files (-filter #'f-exists?
                                    (-concat
                                     (if (my/is-working-hours)
                                         (cdr (assoc 'work my/agenda-files))
                                       (cdr (assoc 'personal my/agenda-files))))))

    (setq org-stuck-projects
          '("+LEVEL=1/-DONE" ("TODO" "NEXT" "NEXTACTION") nil ""))

    ;; https://www.tompurl.com/2015-12-29-emacs-eisenhower-matrix.html
    (setq org-tag-alist '(("important" . ?i)
                          ("urgent" . ?u)))

    (setq org-agenda-custom-commands
          '(("n" "Agenda and all TODOs"
             ((agenda "" ((org-agenda-span 'week)))
              (tags-todo "DEADLINE<=\"<+7d>\""
                         ((org-agenda-overriding-header "Due soon")))
              (todo ""
                    ((org-agenda-overriding-header "Unscheduled")
                     (org-agenda-skip-function '(org-agenda-skip-entry-if 'scheduled 'deadline)))))

             ((org-agenda-start-with-log-mode t)
              (org-agenda-start-day nil)
              (org-agenda-span 'day)
              (org-agenda-log-mode-items '(clock state closed)))
             ("~/Public/org/agenda.html"
              "~/Public/org/agenda.ics"))
            ("l" "Log"
             agenda ""
             ((org-agenda-span 'fortnight)
              (org-agenda-start-day "-1w")
              (org-agenda-start-with-log-mode t)
              (org-agenda-log-mode-items '(clock state closed))
              (org-agenda-include-deadlines nil)
              (org-agenda-skip-scheduled-delay-if-deadline t))
             ("~/Public/org/agenda-log.html"))
            ("e" "Eisenhower Matrix"
             ((tags-todo "+important+urgent"
                         ((org-agenda-overriding-header "Do")))
              (tags-todo "+important-urgent"
                         ((org-agenda-overriding-header "Decide")))
              (tags-todo "-important+urgent"
                         ((org-agenda-overriding-header "Delegate")))
              (tags-todo "-important-urgent"
                         ((org-agenda-overriding-header "Delete"))))
             ((org-agenda-start-with-log-mode t)
              (org-agenda-span 'day)
              (org-agenda-log-mode-items '(clock state closed))))))

    (defun my/org-agenda-timeline ()
      (interactive)
      (let ((org-agenda-files (list (buffer-file-name))))
        (org-agenda)))

    (setq org-agenda-start-on-weekday nil)
    (setq org-agenda-span 'fortnight)
    (setq org-agenda-todo-ignore-scheduled 'future)
    (setq org-agenda-tags-todo-honor-ignore-options t)
    (setq org-agenda-skip-deadline-prewarning-if-scheduled t)
    (add-hook 'org-agenda-finalize-hook (lambda () (hl-line-mode)))

    (setq
     org-icalendar-use-scheduled '(todo-start event-if-todo)
     org-icalendar-combined-agenda-file (expand-file-name "~/Documents/org.ics")))
#+end_src
*** LaTeX Export
**** Shell escaping
Add =-shell-escape= to the latex compilation. Needed for the svg package.
#+begin_src emacs-lisp
  (setq org-latex-pdf-process
        '("%latex -shell-escape -interaction nonstopmode -output-directory %o %f"
          "%latex -shell-escape -interaction nonstopmode -output-directory %o %f"
          "%latex -shell-escape -interaction nonstopmode -output-directory %o %f"))
#+end_src
**** Document Classes
Tell Emacs about all of the LaTeX classes I use to export documents.

#+BEGIN_SRC emacs-lisp
  (use-package! ox-latex
    :config
    (seq-map (apply-partially #'add-to-list 'org-latex-classes)
           '(("koma-letter"
              "\\documentclass{scrlttr2}"
              ("\\section{%s}" . "\\section*{%s}")
              ("\\subsection{%s}" . "\\subsection*{%s}")
              ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
              ("\\paragraph{%s}" . "\\paragraph*{%s}")
              ("\\subparagraph{%s}" . "\\subparagraph*{%s}"))
             ("koma-article"
              "\\documentclass{scrartcl}"
              ("\\section{%s}" . "\\section*{%s}")
              ("\\subsection{%s}" . "\\subsection*{%s}")
              ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
              ("\\paragraph{%s}" . "\\paragraph*{%s}")
              ("\\subparagraph{%s}" . "\\subparagraph*{%s}"))
             ("koma-book"
              "\\documentclass{scrbook}"
              ("\\section{%s}" . "\\section*{%s}")
              ("\\subsection{%s}" . "\\subsection*{%s}")
              ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
              ("\\paragraph{%s}" . "\\paragraph*{%s}")
              ("\\subparagraph{%s}" . "\\subparagraph*{%s}"))
             ("koma-book-chapters"
              "\\documentclass{scrbook}"
              ("\\chapter{%s}" . "\\chapter*{%s}")
              ("\\section{%s}" . "\\section*{%s}")
              ("\\subsection{%s}" . "\\subsection*{%s}")
              ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
              ("\\paragraph{%s}" . "\\paragraph*{%s}")
              ("\\subparagraph{%s}" . "\\subparagraph*{%s}"))
             ("koma-report"
              "\\documentclass{scrreprt}"
              ("\\chapter{%s}" . "\\chapter*{%s}")
              ("\\section{%s}" . "\\section*{%s}")
              ("\\subsection{%s}" . "\\subsection*{%s}")
              ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
              ("\\paragraph{%s}" . "\\paragraph*{%s}")
              ("\\subparagraph{%s}" . "\\subparagraph*{%s}"))
             ("memoir"
              "\\documentclass{memoir}"
              ("\\section{%s}" . "\\section*{%s}")
              ("\\subsection{%s}" . "\\subsection*{%s}")
              ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
              ("\\paragraph{%s}" . "\\paragraph*{%s}")
              ("\\subparagraph{%s}" . "\\subparagraph*{%s}"))
             ("hitec"
              "\\documentclass{hitec}"
              ("\\section{%s}" . "\\section*{%s}")
              ("\\subsection{%s}" . "\\subsection*{%s}")
              ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
              ("\\paragraph{%s}" . "\\paragraph*{%s}")
              ("\\subparagraph{%s}" . "\\subparagraph*{%s}"))
             ("paper"
              "\\documentclass{paper}"
              ("\\section{%s}" . "\\section*{%s}")
              ("\\subsection{%s}" . "\\subsection*{%s}")
              ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
              ("\\paragraph{%s}" . "\\paragraph*{%s}")
              ("\\subparagraph{%s}" . "\\subparagraph*{%s}"))
             ("letter"
              "\\documentclass{letter}"
              ("\\section{%s}" . "\\section*{%s}")
              ("\\subsection{%s}" . "\\subsection*{%s}")
              ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
              ("\\paragraph{%s}" . "\\paragraph*{%s}")
              ("\\subparagraph{%s}" . "\\subparagraph*{%s}"))
             ("tufte-handout"
              "\\documentclass{tufte-handout}"
              ("\\section{%s}" . "\\section*{%s}")
              ("\\subsection{%s}" . "\\subsection*{%s}")
              ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
              ("\\paragraph{%s}" . "\\paragraph*{%s}")
              ("\\subparagraph{%s}" . "\\subparagraph*{%s}"))
             ("tufte-book"
              "\\documentclass{tufte-book}"
              ("\\section{%s}" . "\\section*{%s}")
              ("\\subsection{%s}" . "\\subsection*{%s}")
              ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
              ("\\paragraph{%s}" . "\\paragraph*{%s}")
              ("\\subparagraph{%s}" . "\\subparagraph*{%s}"))
             ("tufte-book-chapters"
              "\\documentclass{tufte-book}"
              ("\\chapter{%s}" . "\\chapter*{%s}")
              ("\\section{%s}" . "\\section*{%s}")
              ("\\subsection{%s}" . "\\subsection*{%s}")
              ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
              ("\\paragraph{%s}" . "\\paragraph*{%s}")
              ("\\subparagraph{%s}" . "\\subparagraph*{%s}"))
             ("labbook"
              "\\documentclass{labbook}"
              ("\\chapter{%s}" . "\\chapter*{%s}")
              ("\\section{%s}" . "\\section*{%s}")
              ("\\subsection{%s}" . "\\labday{%s}")
              ("\\subsubsection{%s}" . "\\experiment{%s}")
              ("\\paragraph{%s}" . "\\paragraph*{%s}")
              ("\\subparagraph{%s}" . "\\subparagraph*{%s}"))
             ("thermal-paper"
              "\\documentclass{paper}
  \\usepackage[paperwidth=52mm]{geometry}"
              ("\\section{%s}" . "\\section*{%s}")
              ("\\subsection{%s}" . "\\subsection*{%s}")
              ("\\subsubsection{%s}" . "\\subsubsection*{%s}")
              ("\\paragraph{%s}" . "\\paragraph*{%s}")
              ("\\subparagraph{%s}" . "\\subparagraph*{%s}")))))
#+END_SRC
**** DnD
This adds an additional LaTeX export option that outputs documents resembling a
Dungeons and Dragons manual.

#+begin_src emacs-lisp
  (use-package! ox-dnd
    :after ox)
#+end_src
*** Capture templates
Set up my capture templates for making new notes and journal entries.

#+begin_src emacs-lisp
  (setq org-capture-templates
        `(
          ;; Personal
          ("j" "Journal Entry" plain
           (file+datetree "~/org/journal.org")
           "%U\n\n%?" :empty-lines-before 1)
          ("t" "TODO" entry
           (file+headline "~/Nextcloud/org/personal.org" "Unsorted")
           "* TODO %^{Description}\n%?")
          ("n" "Note" entry
           (file+headline "~/Nextcloud/org/personal.org" "Notes")
           "* %^{Description}\n%U\n\n%?")
          ("c" "Cookbook Recipe" entry
           (file "~/org/cookbook/index.org")
           "%(org-chef-get-recipe-from-url)"
           :empty-lines 1)
          ;; Org-Protocol
          ("b" "Bookmark" entry
           (file+headline "~/org/bookmarks.org" "Unsorted")
           "* %^{Title}\n\n Source: %u, %c\n\n %i")
          ("p" "Webpage" entry
           (file "~/org/articles.org")
           "* %a\n\n%U %?\n\n%:initial")

          ;; Email
          ;; https://martinralbrecht.wordpress.com/2016/05/30/handling-email-with-emacs/
          ("r" "respond to email (mu4e)"
           entry (file+headline "~/org/todo.org" "Email")
           "* REPLY to [[mailto:%:fromaddress][%:fromname]] on %a\nDEADLINE: %(org-insert-time-stamp (org-read-date nil t \"+1d\"))\n%U\n\n"
           :immediate-finish t
           :prepend t)

          ;; Work
          ;; ("w" "Work")
          ))
#+end_src
*** Custom ID generation
Because I'm all kinds of crazy, I like the custom IDs of my work log entries to
be based on their headings.

#+begin_src emacs-lisp
  (use-package! org-id
    :after org
    :config

    ;; https://writequit.org/articles/emacs-org-mode-generate-ids.html#automating-id-creation
    (defun eos/org-custom-id-get (&optional pom create prefix)
      "Get the CUSTOM_ID property of the entry at point-or-marker POM.
     If POM is nil, refer to the entry at point. If the entry does
     not have an CUSTOM_ID, the function returns nil. However, when
     CREATE is non nil, create a CUSTOM_ID if none is present
     already. PREFIX will be passed through to `org-id-new'. In any
     case, the CUSTOM_ID of the entry is returned."
      (interactive)
      (org-with-point-at pom
        (let ((id (org-entry-get nil "CUSTOM_ID")))
          (cond
           ((and id (stringp id) (string-match "\\S-" id))
            id)
           (create
            (setq id (org-id-new (concat prefix "h")))
            (org-entry-put pom "CUSTOM_ID" id)
            (org-id-add-location id (buffer-file-name (buffer-base-buffer)))
            id)))))

    (defun eos/org-add-ids-to-headlines-in-file ()
      "Add CUSTOM_ID properties to all headlines in the current
     file which do not already have one. Only adds ids if the
     `auto-id' option is set to `t' in the file somewhere. ie,
     ,#+OPTIONS: auto-id:t"
      (interactive)
      (save-excursion
        (widen)
        (goto-char (point-min))
        (when (re-search-forward "^#\\+OPTIONS:.*auto-id:t" (point-max) t)
          (org-map-entries (lambda () (eos/org-id-get (point) 'create)))))
      (save-excursion
        (widen)
        (goto-char (point-min))
        (when (re-search-forward "^#\\+OPTIONS:.*auto-id:worklog" (point-max) t)
          (let ((my/org-worklog-id-depth 2))
            (org-map-entries (lambda () (my/org-worklog-id-get (point) 'create))))))
      (save-excursion
        (widen)
        (goto-char (point-min))
        (when (re-search-forward "^#\\+OPTIONS:.*auto-id:readable" (point-max) t)
          (let ((my/org-worklog-id-depth 0))
            (org-map-entries (lambda () (my/org-worklog-id-get (point) 'create)))))))

    ;; automatically add ids to saved org-mode headlines
    (add-hook 'org-mode-hook
              (lambda ()
                (add-hook 'before-save-hook
                          (lambda ()
                            (when (and (eq major-mode 'org-mode)
                                       (eq buffer-read-only nil))
                              (eos/org-add-ids-to-headlines-in-file))))))

    (defun my/org-remove-all-ids ()
      (interactive)
      (save-excursion
        (widen)
        (goto-char (point-min))

.config/qtile/config.py
import dataclasses
import datetime
import os
import pathlib
import platform
import random
import shutil
import subprocess
import typing

import libqtile.resources
from libqtile import bar, hook, layout, qtile, widget, utils
from libqtile.config import Click, Drag, Group, Key, Match, Screen
from libqtile.lazy import lazy
from libqtile.widget import backlight
from libqtile.backend.wayland import InputConfig


def first[T](
    options: list[T],
    predicate: typing.Callable[[T], typing.Any] | None = None,
    default: T | None = None,
) -> T:
    predicate = predicate or (lambda _: True)
    try:
        return next(option for option in options if predicate(option))
    except StopIteration:
        if default:
            return default
        else:
            raise RuntimeError("No valid options available")


def wp_path():
    wp_path = pathlib.Path(os.path.expanduser("~/Pictures/Wallpapers"))
    if qtile.current_screen.width > 4000:
        wp_path = wp_path / "Ultrawide"
    return wp_path


def random_wallpaper():
    wallpapers = [
        f
        for f in wp_path().iterdir()
        if f.is_file() and f.suffix.lower() in [".png", ".jpg", ".jpeg"]
    ]
    return random.choice(wallpapers)


@dataclasses.dataclass
class TimeRange:
    start: datetime.time
    end: datetime.time
    dow: int

    def contains(self, check_time: datetime.datetime) -> bool:
        return all(
            [
                check_time.time() >= self.start,
                check_time.time() <= self.end,
                check_time.weekday() == self.dow,
            ]
        )


workdays = [
    # Monday (0) through Friday (4)
    TimeRange(datetime.time(8), datetime.time(18), dow)
    for dow in [0, 1, 2, 3, 4]
]
# Sunday (6)
workdays.append(TimeRange(datetime.time(12), datetime.time(15), 6))
workdays.append(TimeRange(datetime.time(0), datetime.time(15), 5))


def during_work_hours(check_time: datetime.datetime | None = None) -> bool:
    if not check_time:
        check_time = datetime.datetime.now()
    return any([workday.contains(check_time) for workday in workdays])


def preferred_wallpaper(screen):
    wp_sunny = ("wide", "safe", "~/Pictures/Wallpapers/Ultrawide/o7i7rtkdwyd91.jpg")
    wp_lapis = ("full", "nsfw", "~/Pictures/Wallpapers/20210704_191749.jpeg")
    wp_cloud = ("full", "safe", "~/Pictures/Wallpapers/cloud_by_sakimichan-d4ly1t6.jpg")
    wp_xenia = ("full", "safe", "~/Pictures/Wallpapers/Xenia.png")
    preferences = {
        "framework13": [wp_sunny, wp_lapis, wp_cloud],
        "snappy": [wp_sunny, wp_xenia],
    }
    size = "wide" if qtile and qtile.current_screen.width > 4000 else "full"
    safety = "safe" if during_work_hours() else "nsfw"
    wallpapers = preferences.get(platform.node(), "framework13")

    def predicate(wp) -> bool:
        wp_size, wp_safety, wp_path = wp
        return wp_size == size and (wp_safety == safety or wp_safety == "safe")

    wallpaper = first(wallpapers, predicate, wallpapers[-1])
    return os.path.expanduser(wallpaper[-1])


def set_random_wallpaper(screen):
    screen.set_wallpaper(random_wallpaper(), "fill")


def set_preferred_wallpaper(screen):
    screen.set_wallpaper(preferred_wallpaper(screen), "fill")


mod = "mod4"
terminal = first(["konsole", "alacritty", "xterm"], shutil.which, "konsole")
volume = os.path.expanduser("~/.config/qtile/dunst-volume.sh")
keys = [
    # A list of available commands that can be bound to keys can be found
    # at https://docs.qtile.org/en/latest/manual/config/lazy.html
    # Switch between windows
    Key([mod], "h", lazy.layout.left(), desc="Move focus to left"),
    Key([mod], "l", lazy.layout.right(), desc="Move focus to right"),
    Key([mod], "j", lazy.layout.down(), desc="Move focus down"),
    Key([mod], "k", lazy.layout.up(), desc="Move focus up"),
    # Move windows between left/right columns or move up/down in current stack.
    # Moving out of range in Columns layout will create new column.
    Key(
        [mod, "shift"], "h", lazy.layout.shuffle_left(), desc="Move window to the left"
    ),
    Key(
        [mod, "shift"],
        "l",
        lazy.layout.shuffle_right(),
        desc="Move window to the right",
    ),
    Key([mod, "shift"], "j", lazy.layout.shuffle_down(), desc="Move window down"),
    Key([mod, "shift"], "k", lazy.layout.shuffle_up(), desc="Move window up"),
    # Grow windows. If current window is on the edge of screen and direction
    # will be to screen edge - window would shrink.
    Key([mod, "control"], "h", lazy.layout.grow_left(), desc="Grow window to the left"),
    Key(
        [mod, "control"], "l", lazy.layout.grow_right(), desc="Grow window to the right"
    ),
    Key([mod, "control"], "j", lazy.layout.grow_down(), desc="Grow window down"),
    Key([mod, "control"], "k", lazy.layout.grow_up(), desc="Grow window up"),
    Key([mod], "n", lazy.layout.normalize(), desc="Reset all window sizes"),
    # Toggle between split and unsplit sides of stack.
    # Split = all windows displayed
    # Unsplit = 1 window displayed, like Max layout, but still with
    # multiple stack panes
    Key(
        [mod, "shift"],
        "Return",
        lazy.layout.toggle_split(),
        desc="Toggle between split and unsplit sides of stack",
    ),
    Key([mod], "Return", lazy.spawn(terminal), desc="Launch terminal"),
    # Toggle between different layouts as defined below
    Key([mod], "Tab", lazy.next_layout(), desc="Toggle between layouts"),
    Key([mod], "w", lazy.window.kill(), desc="Kill focused window"),
    Key(
        [mod],
        "f",
        lazy.window.toggle_fullscreen(),
        desc="Toggle fullscreen on the focused window",
    ),
    Key(
        [mod],
        "t",
        lazy.window.toggle_floating(),
        desc="Toggle floating on the focused window",
    ),
    Key([mod, "control"], "r", lazy.reload_config(), desc="Reload the config"),
    Key([mod, "control"], "q", lazy.shutdown(), desc="Shutdown Qtile"),
    Key([mod], "space", lazy.spawn("rofi -show drun")),
    Key([], "XF86AudioRaiseVolume", lazy.spawn(f"{volume} +2%")),
    Key([], "XF86AudioLowerVolume", lazy.spawn(f"{volume} -2%")),
    Key(
        [],
        "XF86MonBrightnessUp",
        lazy.widget["backlight"].change_backlight(backlight.ChangeDirection.UP),
    ),
    Key(
        [],
        "XF86MonBrightnessDown",
        lazy.widget["backlight"].change_backlight(backlight.ChangeDirection.DOWN),
    ),
    Key([], "Print", lazy.spawn("spectacle")),
    Key([mod, "control"], "w", lazy.screen.function(set_random_wallpaper)),
    Key([mod, "shift"], "w", lazy.screen.function(set_preferred_wallpaper)),
    Key([mod, "shift", "control"], "l", lazy.spawn("slock")),
    # Somehow, previous and next are backwards?
    Key([mod], "comma", lazy.next_screen()),
    Key([mod], "period", lazy.prev_screen()),
    Key([mod], "p", lazy.spawn("rofi-power")),
    Key([mod], "s", lazy.spawn("rofi-sound")),
    Key([mod], "m", lazy.spawn("xfce4-find-cursor")),
]

# Add key bindings to switch VTs in Wayland.
# We can't check qtile.core.name in default config as it is loaded before qtile is started
# We therefore defer the check until the key binding is run by using .when(func=...)
for vt in range(1, 8):
    keys.append(
        Key(
            ["control", "mod1"],
            f"f{vt}",
            lazy.core.change_vt(vt).when(func=lambda: qtile.core.name == "wayland"),
            desc=f"Switch to VT{vt}",
        )
    )


groups = [Group(i) for i in "123456789"]

for i in groups:
    keys.extend(
        [
            # mod + group number = switch to group
            Key(
                [mod],
                i.name,
                lazy.group[i.name].toscreen(),
                desc=f"Switch to group {i.name}",
            ),
            # mod + shift + group number = switch to & move focused window to group
            # Key(
            #     [mod, "shift"],
            #     i.name,
            #     lazy.window.togroup(i.name, switch_group=True),
            #     desc=f"Switch to & move focused window to group {i.name}",
            # ),
            # Or, use below if you prefer not to switch to that group.
            # mod + shift + group number = move focused window to group
            Key(
                [mod, "shift"],
                i.name,
                lazy.window.togroup(i.name),
                desc="move focused window to group {}".format(i.name),
            ),
        ]
    )

layout_theme = {
    "border_width": 5,
    "margin": 30,
    "border_focus": "#d75f5f",
    "border_normal": "#8f3d3d",
    "single_border_width": 5,
}

layouts = [
    # layout.Columns(border_focus_stack=["#d75f5f", "#8f3d3d"], border_width=4),
    layout.MonadThreeCol(**layout_theme),
    layout.Columns(**layout_theme),
    layout.VerticalTile(**layout_theme),
    layout.TreeTab(**layout_theme),
    # layout.Max(),
    # Try more layouts by unleashing below layouts.
    # layout.Stack(num_stacks=2),
    # layout.Bsp(),
    # layout.Matrix(),
    # layout.MonadTall(),
    # layout.MonadWide(),
    # layout.RatioTile(),
    # layout.Tile(),
    # layout.TreeTab(),
    # layout.VerticalTile(),
    # layout.Zoomy(),
]

widget_defaults = dict(
    font="Beleren",
    fontsize=18,
    padding=3,
)
extension_defaults = widget_defaults.copy()

logo = os.path.join(os.path.dirname(libqtile.resources.__file__), "logo.png")
screens = [
    Screen(
        top=bar.Bar(
            [
                widget.CurrentLayout(),
                widget.GroupBox(),
                widget.Spacer(),
                widget.WindowName(),
                widget.Spacer(),
                widget.DoNotDisturb(disabled_icon="📣", enabled_icon="😴"),
                widget.Backlight(),
                # NB Systray is incompatible with Wayland, consider using StatusNotifier instead
                widget.StatusNotifier(),
                widget.Systray(hide_crash=True),
                widget.BatteryIcon(),
                widget.Battery(),
                widget.Clock(format="%Y-%m-%d %a %I:%M %p"),
                # widget.QuickExit(),
            ],
            28,
            # border_width=[2, 0, 2, 0],  # Draw top and bottom borders
            # border_color=["ff00ff", "000000", "ff00ff", "000000"]  # Borders are magenta
        ),
        background="#000000",
        # wallpaper=logo,
        # wallpaper_mode="fill",
        # You can uncomment this variable if you see that on X11 floating resize/moving is laggy
        # By default we handle these events delayed to already improve performance, however your system might still be struggling
        # This variable is set to None (no cap) by default, but you can set it to 60 to indicate that you limit it to 60 events per second
        # x11_drag_polling_rate = 60,
    )
    for _ in range(3)
]

# Drag floating layouts.
mouse = [
    Drag(
        [mod],
        "Button1",
        lazy.window.set_position_floating(),
        start=lazy.window.get_position(),
    ),
    Drag(
        [mod], "Button3", lazy.window.set_size_floating(), start=lazy.window.get_size()
    ),
    Click([mod], "Button2", lazy.window.bring_to_front()),
]

dgroups_key_binder = None
dgroups_app_rules = []  # type: list
follow_mouse_focus = True
bring_front_click = False
floats_kept_above = True
cursor_warp = False
floating_layout = layout.Floating(
    float_rules=[
        # Run the utility of `xprop` to see the wm class and name of an X client.
        *layout.Floating.default_float_rules,
        Match(wm_class="confirmreset"),  # gitk
        Match(wm_class="makebranch"),  # gitk
        Match(wm_class="maketag"),  # gitk
        Match(wm_class="ssh-askpass"),  # ssh-askpass
        Match(title="branchdialog"),  # gitk
        Match(title="pinentry"),  # GPG key password entry
    ]
)
auto_fullscreen = True
focus_on_window_activation = "smart"
focus_previous_on_window_remove = False
reconfigure_screens = True

# If things like steam games want to auto-minimize themselves when losing
# focus, should we respect this or not?
auto_minimize = True

# When using the Wayland backend, this can be used to configure input devices.
wl_input_rules = {
    "type:keyboard": InputConfig(kb_options="ctrl:nocaps"),
}

# xcursor theme (string or None) and size (integer) for Wayland backend
wl_xcursor_theme = None
wl_xcursor_size = 24

idle_inhibitors = []  # type: list

# XXX: Gasp! We're lying here. In fact, nobody really uses or cares about this
# string besides java UI toolkits; you can see several discussions on the
# mailing lists, GitHub issues, and other WM documentation that suggest setting
# this string if your java app doesn't work correctly. We may as well just lie
# and say that we're a working one by default.
#
# We choose LG3D to maximize irony: it is a 3D non-reparenting WM written in
# java that happens to be on java's whitelist.
wmname = "LG3D"


@hook.subscribe.startup
@hook.subscribe.screens_reconfigured
def screen_reconf():
    utils.send_notification("qtile", "Screens reconfigured")
    for screen in qtile.screens:
        set_preferred_wallpaper(screen)


@hook.subscribe.startup_once
def autostart():
    script = os.path.expanduser("~/.config/qtile/autostart.sh")
    subprocess.call([script])
